ui/src/coordinator.rs

```diff
@@ -0,0 +1,1284 @@
+                }
+            },
+            worker_msg = worker_msg_receiver.recv() => {
+                if let Some(msg) = worker_msg {
+                    match msg {
+                        WorkerMessage::Response(id, resp) => {
+                            if let Some(tx) = worker_response_senders.remove(&id) {
+                                tx.send(resp).ok().context(UnableToSendJobReportSnafu)?;
+                            }
+                        }
+                        WorkerMessage::StdoutPacket(packet) => {
+                            let Ok(_) = ws_sender.send(
+                                serde_json::to_string(&PlaygroundMessage::Stdout {packet}).context(PlaygroundMessageSerializationSnafu)?).await
+                                .context(UnableToSendPlaygroundMessageToWebSocketSnafu)
+                            else {
+                                // TODO: log the failure.
+                                continue;
+                            };
+                        }
+                        WorkerMessage::StderrPacket(packet) => {
+                            let Ok(_) = ws_sender.send(
+                                serde_json::to_string(&PlaygroundMessage::Stderr {packet}).context(PlaygroundMessageSerializationSnafu)?).await
+                                .context(UnableToSendPlaygroundMessageToWebSocketSnafu)
+                            else {
+                                // TODO: log the failure.
+                                continue;
+                            };
+                        }
+                        WorkerMessage::StreamFailure => {
+                            let Ok(_) = ws_sender.send(
+                                serde_json::to_string(&PlaygroundMessage::StreamFailure).context(PlaygroundMessageSerializationSnafu)?).await
+                                .context(UnableToSendPlaygroundMessageToWebSocketSnafu)
+                            else {
+                                // TODO: log the failure.
+                                continue;
+                            };
+                        }
+                    }
+                } else {
+                    // Shouldn't be closed unless worker panicked or died.
+                    return Err(Error::UnableToReceiveWorkerMessage);
+                }
+            }
+        }
+    }
+}
+
+pub fn spawn_worker() -> Result<Container> {
+    let mut tasks = JoinSet::new();
+    let (stdin, stdout) = run_worker_in_background()?;
+    let (coordinator_msg_sender, worker_msg_receiver) = spawn_io_queue(&mut tasks, stdin, stdout);
+    let (client_tx, ws_receiver) = mpsc::channel(8);
+    let (ws_sender, client_rx) = mpsc::channel(8);
+    tasks.spawn(async move {
+        pair_websocket_worker(
+            ws_sender,
+            ws_receiver,
+            coordinator_msg_sender,
+            worker_msg_receiver,
+        )
+        .await
+    });
+    tokio::spawn(async move {
```

Comment:
I'm not a fan of spawning tasks and never checking them again; I prefer to put them in a `JoinSet`.